programs/host/frontend.c

```diff
@@ -0,0 +1,937 @@
+            break;
+        }
+    }
+    return( file_id );
+}
+
+static int free_file_context( int32_t file_id )
+{
+    int ret = -1;
+    file_id--;
+    if ( file_id < MBEDTLS_SERIALIZE_MAX_FILES )
+    {
+        files[file_id].inUse = 0;
+        if ( files[file_id].file )
+        {
+            files[file_id].file = NULL;
+            ret = 0;
+        }
+    }
+    return( ret );
+}
+
+static void * get_file_from_id( int32_t file_id )
+{
+    file_id--;
+    if ( file_id < MBEDTLS_SERIALIZE_MAX_FILES && files[file_id].inUse )
+    {
+        return files[file_id].file;
+    }
+    return( NULL );
+}
+
+/** Execute an offloaded function.
+ *
+ * @param ctx       Offloading context
+ * @param function  Function to execute (MBEDTLS_SERIALIZE_FUNCTION_xxx)
+ * @param outputs   Array[16] of outputs. Initially all-null. On return,
+ *                  for a function with N outputs, positions 0 to N-1 will
+ *                  be filled with output parameters allocated with alloc_item.
+ * @return status   0 for ok, MBEDTLS_ERR_xxx on error
+ */
+static uint32_t mbedtls_serialize_perform( mbedtls_serialize_context_t *ctx_p,
+                                           uint32_t function,
+                                           mbedtls_serialize_item_t **outputs )
+{
+    mbedtls_serialize_item_t *inputs[16] = {0};
+    mbedtls_serialize_item_t *item;
+    size_t arity;
+    int ret = 0;
+
+    for( arity = 0, item = ctx_p->stack;
+         arity < ( ( function & 0x0000f0 ) >> 4 ) && item != NULL;
+         arity++, item = item->next )
+    {
+        inputs[arity] = item;
+    }
+    DBG( "arity=%zu", arity );
+
+    switch( function )
+    {
+        case MBEDTLS_SERIALIZE_FUNCTION_EXIT:
+            ret = 0;
+            ctx_p->status = MBEDTLS_SERIALIZE_STATUS_EXITED;
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_ECHO:
+            {
+                CHECK_ARITY( 1 );
+                ALLOC_OUTPUT( 0, inputs[0]->size );
+                DBG( "executing echo" );
+                memcpy( item_buffer( outputs[0] ), item_buffer( inputs[0] ),
+                        inputs[0]->size );
+                ret = 0;
+                break;
+            }
+
+        case MBEDTLS_SERIALIZE_FUNCTION_USLEEP:
+            CHECK_ARITY( 1 );
+            CHECK_LENGTH( 0, 4 ); // usec
+            {
+                unsigned long usec = item_uint32( inputs[0] );
+                DBG( "executing sleep usec=%lu", usec );
+                mbedtls_net_usleep( usec );
+                ret = 0;
+            }
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_SOCKET:
+            CHECK_ARITY( 3 ); // host, port, proto_and_mode
+            CHECK_LENGTH( 2, 2 ); // proto_and_mode
+            {
+                char *host = item_buffer( inputs[0] );
+                char *port = item_buffer( inputs[1] );
+                if ( host[inputs[0]->size - 1] == '\0' && port[inputs[1]->size - 1] == '\0' )
```

Comment:
Should we add a length check for `inputs[0]` before making this access?

Reply:
Control will not reach this stage if the input buffer size is 0. It will fail while trying to receive 0 length data. However, there should be checks to report bad input.